File: SinglyLinkedList/list.cpp
#include <iostream>
using namespace std;

class node
{
public:
	node(int d, node* n) : data(d), next(n) {}
	int getdata() { return data; }
	void setdata(int data) { this->data = data; }
	node* getnext() { return next; }
	void setnext(node* next) { this->next = next; }

private:
	int data;
	node* next;
};


class list
{
public:
	list() { head = nullptr; listSize = 0; }

	int size() { return listSize; }
	bool empty() { return size() == 0; }
	void push_front(int data)
	{
		node* newNode = new node(data, nullptr);
		newNode->setnext(head);
		head = newNode;
		++listSize;
	}

	void pop_front()
	{
		if(empty())
		{
			cout << "Cannot pop from empty list\n";
			return;
		}
		node* temp = head;
		head = head->getnext();
		delete temp;
		--listSize;
	}

	void push_back(int data)
	{
		node* newNode = new node(data, nullptr);
		if(empty())
		{
			push_front(data);
			return;
		}

		node* it = head;
		while(it->getnext() != nullptr)
		{
			it = it->getnext();
		}

		it->setnext(newNode);
		++listSize;
	}

	void pop_back()
	{
		if(size() <= 1)
		{
			pop_front();
			return;
		}

		node* it = head;
		node* trailer = nullptr;

		while(it->getnext() != nullptr)
		{
			trailer = it;
			it = it->getnext();
		}
		trailer->setnext(nullptr);
		delete it;
		--listSize;
	}

	void print()
	{
		node* it = head;
		while(it != nullptr)
		{
			cout << it->getdata() << " -> ";
			it = it->getnext();
		}
		cout << "null\n";
	}

	void reverseList()
	{
		reverseListHelper(head, nullptr);
	}

	void clear()
	{
		while(!empty())
		{
			pop_front();
		}
	}

	~list()
	{
		clear();
	}


	

private:
	node* head;
	int listSize;

	void reverseListHelper(node* lead, node* tail)
	{
		if(lead == nullptr)
		{
			head = tail;
			return;
		}

		reverseListHelper(lead->getnext(), lead);

		lead->setnext(tail);

	}

};



int main()
{
	list myList;

	cout << "Size: " << myList.size() << endl;
	cout << "Empty: " << myList.empty() << endl;
	cout << "------------------------------\n";

	for(int i = 1; i < 11; ++i)
	{
		myList.push_front(i);
	}

	myList.print();
	cout << "Size: " << myList.size() << endl;
	cout << "Empty: " << myList.empty() << endl;

	cout << "Popping front\n";
	myList.pop_front();
	cout << "Popping front\n";
	myList.pop_front();
	cout << "Popping front\n";
	myList.pop_front();

	cout << "Popping back\n";
	myList.pop_back();
	cout << "Popping back\n";
	myList.pop_back();
	cout << "Popping back\n";
	myList.pop_back();

	cout << "------------------------------\n";
	myList.print();
	cout << "Size: " << myList.size() << endl;
	cout << "Empty: " << myList.empty() << endl;
	cout << "------------------------------\n";

	for(int i = 0; i < 6; ++i)
	{
		myList.push_back(i);
	}

	myList.print();
	cout << "Size: " << myList.size() << endl;
	cout << "Empty: " << myList.empty() << endl;
	cout << "------------------------------\n";


	cout << "Popping front\n";
	myList.pop_front();
	cout << "Popping front\n";
	myList.pop_front();
	cout << "Popping front\n";
	myList.pop_front();
	cout << "Popping front\n";
	myList.pop_front();
	cout << "------------------------------\n";

	myList.print();
	cout << "Size: " << myList.size() << endl;
	cout << "Empty: " << myList.empty() << endl;
	myList.clear();

	cout << "------------------------------\n";
	for(int i = 1; i < 11; ++i)
	{
		myList.push_front(i);
	}
	myList.print();
	cout << "Calling reverseList...\n";
	myList.reverseList();
	myList.print();


	return 0;
}

File: LinkedListinPython/list.py
# implementing a linked list

class node(object):
	def __init__(self, next=None, data=None):
		self.next = next
		self.data = data
	
	def getNext(self):
		return self.next
	
	def setNext(self, newNext):
		self.next = newNext
		
	def getData(self):
		return self.data
		
	def setData(self, newData):
		self.data = newData
		
	def delete(self):
		del self.data
		
class list(object):
	def __init__(self):
		self.head = None
		self.size = 0
	
	def list_size(self):
		return self.size
	
	def push_front(self, newData=None):
		# testing if newData is None
		if newData is None:
			print("No data received..")
			return
		
		newNode = node(None, newData)
		if self.size == 0:
			self.head = newNode
			self.size += 1
			return
			
		newNode.setNext(self.head)
		self.head = newNode
		self.size += 1
		return
		
	def front(self):
		return self.head.getData()
		
	def push_back(self, newData):
		if self.list_size() == 0:
			self.push_front(newData)
			return
		
		newNode = node(None, newData)
		it = self.head
		while it.getNext() is not None:
			it = it.getNext()
		it.setNext(newNode)
		self.size += 1
		return
	
	def display(self):
		print("There are {} elements in this list".format(self.list_size()))
		it = self.head
		while it is not None:
			print(str(it.getData()),end=" -> ")
			it = it.getNext()
		print("null\n")
		
	# does not work yet..
'''	def clear(self):
		tracker = self.head
		trailer = None
		
		while tracker:
			trailer = tracker
			tracker = tracker.getNext()
			trailer.delete()
		self.size = 0
'''
		
def main():
	myList = list()
	
	# testing push_front
	myList.push_front(1)
	#testing display
	myList.display()
	
	# testing push_front in a loop
	for i in range(2,10):
		myList.push_front(i)
	myList.display()
	
	# testing front function
	print("The front of the list is:",myList.front(),end="/n/n")

	# testing push_back function
	myList.push_back(0)
	myList.display()

	'''
	# test clear
	myList.clear()
	myList.display()
	'''
	
	# creates a new list and tests push_back with a loop
	newList = list()
	for i in range(0,10):
		newList.push_back(i)
	newList.display()
	
	
	
main()